csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1382,6 +1383,50 @@ class AllocationReuseModifier : private kir::ExprMutator {
   std::unordered_map<kir::Allocate*, kir::Allocate*> old2new_;
 };
 
+//! Set addresses without any re-use of shared memory. This simply scans
+//! through exprs and each time we find an unaliased Allocate align the current
+//! address and increment it by the given amount.
+class NoReuseSharedMemAllocator : kir::IrVisitor {
+ public:
+  void allocate(std::vector<Expr*>& exprs) {
+    handle(exprs);
+  }
+
+ private:
+  using kir::IrVisitor::handle;
+
+  static Val* alignExpr(Val* addr, int64_t alignment = 16) {
+    if (alignment == 1) {
+      return addr;
+    }
+    auto n_minus_one = IrBuilder::create<Val>(alignment - 1);
+    return bitwise_and(add(addr, n_minus_one), bitwise_not(n_minus_one));
```

Comment:
Could you use `IrBuilder::bitwiseAndExpr` instead of `bitwise_and`? Unfortunately our functions in `arith.h` does not support `DataType::Index`, and I have an ongoing PR that changes most indexing vals to `DataType::Index`. Same thing applies to all ops from `arith.h`.

Reply:
Definitely should use `bitwiseAndExpr`. Thanks.

I think the bitwise ops in `arith.cpp` work with `DataType::Index` since they just check `isIntegralType` which should be true for `Index`. But I haven't tried.